cmd/catalyst-node/catalyst-node.go

```diff
@@ -0,0 +1,284 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	serfclient "github.com/hashicorp/serf/client"
+	"github.com/hashicorp/serf/cmd/serf/command/agent"
+	"github.com/mitchellh/cli"
+	"github.com/peterbourgon/ff/v3"
+)
+
+type (
+	catalystConfig struct {
+		serfRPCAddress           string
+		serfRPCAuthKey           string
+		mistLoadBalancerEndpoint string
+	}
+)
+
+var Commands map[string]cli.CommandFactory
+
+func init() {
+	ui := &cli.BasicUi{Writer: os.Stdout}
+
+	// eli note: this is copied from here:
+	// https://github.com/hashicorp/serf/blob/a2bba5676d6e37953715ea10e583843793a0c507/cmd/serf/commands.go#L20-L25
+	// but we should someday get a little bit smarter and invoke serf directly
+	// instead of wrapping their CLI helper
+
+	Commands = map[string]cli.CommandFactory{
+		"agent": func() (cli.Command, error) {
+			a := &agent.Command{
+				Ui:         ui,
+				ShutdownCh: make(chan struct{}),
+			}
+			return a, nil
+		},
+	}
+}
+
+func runClient(config catalystConfig) error {
+
+	client, err := connectSerfAgent(config.serfRPCAddress, config.serfRPCAuthKey)
+
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	eventCh := make(chan map[string]interface{}, 1024)
+	streamHandle, err := client.Stream("*", eventCh)
+	if err != nil {
+		return fmt.Errorf("error starting stream: %w", err)
+	}
+	defer client.Stop(streamHandle)
+
+	// eli note: not sure if this is useful yet, but we can do this as well:
+
+	// logCh := make(chan string, 1024)
+	// monHandle, err := client.Monitor(logutils.LogLevel("INFO"), logCh)
+	// if err != nil {
+	// 	return fmt.Errorf("error starting monitor: %s", err)
+	// }
+	// defer client.Stop(monHandle)
+
+	// eli note: uncertain how we handle dis/reconnects here. but it's local, so hopefully rare?
+	event := <-eventCh
+	inbox := make(chan map[string]interface{}, 1)
+	go func() {
+		for {
+			e := <-eventCh
```

Comment:
What is the logic here? I can't grasp it?

Reply:
Yeah, we had a whole Discord discussion about this pattern - take a look here, and then at the thread under Yondon's post https://discord.com/channels/423160867534929930/445623681251672064/986702686407761920